### PR TITLE
[watermarking] add context deduplication support to speculative decoding

### PR DESCRIPTION
**Pull request by TQCB**
## Purpose

Extend generation-side watermark context deduplication to speculative decoding.

- Build target verification contexts from committed and proposed draft tokens.
- Detect repeated contexts in committed history and earlier positions of the same speculative block while respecting `deduplicate_contexts_max_history`.
- Use the ordinary draft proposal at repeated contexts before running the unchanged acceptance test.
- Use the ordinary draw from the existing residual or target distribution for repeated rejection-recovery and bonus contexts.
- Leave disabled and greedy requests on their existing paths.
- Fuse committed-history and speculative-local context checks to avoid launch-heavy PyTorch bookkeeping.
- Fuse ordinary and watermarked draft sampling into one vocabulary reduction while preserving the pre-temperature draft-logit cache.

This keeps the standard `min(1, q/p)` acceptance rule and does not force rejection at repeated contexts.

## Test Plan

- Run the focused CPU watermark/configuration tests.
- Run all pre-commit hooks applicable to the changed files.
- Run the watermark suite and CPU/Triton parity tests on an H100.
- Microbenchmark draft, target-mask, rejection-resampling, and fused draft-sampling overhead separately.

## Test Result

- `python -m pytest -q tests/watermarking/test_watermarking.py`: 51 passed, 18 skipped locally.
- Watermarking and Gumbel suites on an H100: 95 passed.
- Fused draft sampling exactly matches the previous ordinary/watermarked selection, including draft-specific RNG streams and bit-exact pre-temperature logits caching.
- Applicable Ruff, formatting, syntax, mypy, configuration, accelerator API, and sign-off checks: passed.

### H100 microbenchmark

The new fused draft-sampling path was compared with the previous sequence—ordinary draft sampling with cache write, watermarked sampling, then `torch.where`. Both paths include temperature preprocessing and draft-logit cache writes, use float16 logits with vocabulary size 32768, and report the median of five 1000-iteration measurements.

| Batch size | Previous two-pass path | Fused mixed path | Time saved per draft step | Sampling-section speedup |
| ---: | ---: | ---: | ---: | ---: |
| 1 | 91.31 us | 57.40 us | 33.91 us | 1.59x |
| 8 | 92.96 us | 58.14 us | 34.82 us | 1.60x |
| 32 | 92.62 us | 58.13 us | 34.49 us | 1.59x |

These are isolated eager-execution microbenchmark timings, not end-to-end inference throughput measurements.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://docs.vllm.ai/en/latest/contributing>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/watermark.py [hunk 1]
```diff
@@ -177,10 +299,46 @@ def repeated_context_mask(
...
+                repeated |= (contexts < 0).any(dim=-1)
+        return repeated
 
     if contexts.stride(-1) != 1:
         contexts = contexts.contiguous()
+    # The kernel reads these flat, one element per row. `history_offsets` is
+    # not in the list because the kernel reads it as
+    # `history_offsets_ptr + row * history_offsets_stride`, so a stride-0
+    # broadcast row needs no copy. `local_positions` is read flat and is packed
+    # even when the caller passes the same tensor for both.
```
Comment (simon-veitner-redhat) [inline vllm/v1/worker/gpu/sample/watermark.py]: shorten this

File: vllm/v1/worker/gpu/sample/watermark.py [hunk 2]
```diff
@@ -189,17 +343,109 @@ def repeated_context_mask(
...
+    if contexts.stride(-1) != 1:
+        contexts = contexts.contiguous()
+    # The kernel indexes these by row without a stride. `steps` is left alone:
+    # the kernel takes its stride, so a broadcast row needs no copy.
+    req_indices = req_indices.contiguous()
+    enabled = enabled.contiguous()
+    # `prompt_lens`, `total_lens` and the last dimension of `all_token_ids` and
+    # of `prior_contexts` are read flat and contiguous by contract:
+    # `prior_contexts` is allocated by the draft watermarker and the other
+    # three come from the runner's request state, which owns them outright.
```
Comment (simon-veitner-redhat) [inline vllm/v1/worker/gpu/sample/watermark.py:416]: shorten this

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py [hunk 3]
```diff
@@ -147,14 +158,6 @@ def _sample_logits(
             logits_cache_col=self._step_cols[step],
             use_fp64=self.use_fp64_gumbel,
         )
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu/spec_decode/dspark/speculator.py:160]: could we unify this to 
```
sampler =  gumbel_sample if self.draft_watermarker is None else self.draft_watermarker.sample

sampler(
    ...
)
```
?

File: vllm/v1/worker/gpu/model_runner.py [hunk 4]
```diff
@@ -2143,6 +2146,9 @@ def sample_tokens(
                 self.speculator.prepare_watermarking(
                     self.sampler._get_contexts(input_batch.idx_mapping),
                     self.sampler.watermarking.gpu[input_batch.idx_mapping],
+                    self.sampler.req_states.all_token_ids.gpu,
+                    self.sampler.req_states.prompt_len.gpu,
+                    self.sampler.req_states.total_len.gpu,
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu/model_runner.py:2151]: could we do 

```
    def prepare_watermarking(
        self, sampler: "GPUWatermarkSampler", idx_mapping: torch.Tensor
    )
```
instead? just to simplify this interface / the model runner a bit
